File: app/src/main/java/net/emilla/struct/sort/Unfilter.java
package net.emilla.struct.sort;

import net.emilla.struct.IndexedStruct;

import java.util.Iterator;
import java.util.stream.Stream;

public final class Unfilter<E> implements FilterResult<E> {

    private final IndexedStruct<E> mData;

    public Unfilter(IndexedStruct<E> data) {
        mData = data;
    }

    @Override
    public boolean onePreferredMatch() {
        return size() == 1;
    }

    @Override
    public E get(int index) {
        return mData.get(index);
    }

    @Override
    public int size() {
        return mData.size();
    }

    @Override
    public boolean isEmpty() {
        return mData.isEmpty();
    }

    @Override
    public Stream<E> stream() {
        return mData.stream();
    }

    @Override
    public Iterator<E> iterator() {
        return mData.iterator();
    }

}

File: app/src/main/java/net/emilla/struct/sort/Unfilter.kt
package net.emilla.struct.sort

import androidx.recyclerview.widget.RecyclerView
import net.emilla.view.notifyItemsClamped
import net.emilla.view.notifyTrimmedUntil

internal class Unfilter<E : Searchable<E>>(private val data: SearchableArray<E>) : FilterResult<E> {
    override fun get(index: Int): E = data[index]
    override fun size() = data.size()
    override fun isEmpty() = data.isEmpty()

    override fun narrow(prefixedSearch: String): SearchResult<E> = data.filter(prefixedSearch)

    override fun update(adapter: RecyclerView.Adapter<*>, newFilter: FilterResult<E>) {
        when (newFilter) {
            is EmptyFilter<E> -> {
                adapter.notifyTrimmedUntil(size())
            }
            is WindowSearch<E> -> {
                adapter.notifyItemsClamped(newFilter.window(), size())
            }
            is SparseSearch<E> -> {
                adapter.notifyDataSetChanged()
            }
            is SectoredSearch<E> -> {
                adapter.notifyDataSetChanged()
            }
            is Unfilter<E> -> { /* nothing to do. */ }
        }
    }

    override fun onePreferredMatch() = data.size() == 1
}
